servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpPayloadDiscardCleanerServiceFilter.java

```diff
@@ -0,0 +1,136 @@
+     */
+    static final StreamingHttpServiceFilterFactory INSTANCE = new HttpPayloadDiscardCleanerServiceFilter();
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(HttpPayloadDiscardCleanerServiceFilter.class);
+
+    private HttpPayloadDiscardCleanerServiceFilter() {
+        // Singleton
+    }
+
+    @Override
+    public StreamingHttpServiceFilter create(final StreamingHttpService service) {
+
+        final WatchdogHttpLifecycleObserver observer = new WatchdogHttpLifecycleObserver(this, false);
+
+        return new StreamingHttpServiceFilter(service) {
+            @Override
+            public Single<StreamingHttpResponse> handle(final HttpServiceContext ctx,
+                                                        final StreamingHttpRequest request,
+                                                        final StreamingHttpResponseFactory responseFactory) {
+                return observer.trackLifecycle(ctx, request, r -> delegate().handle(ctx, r, responseFactory));
+            }
+
+            @Override
+            public HttpExecutionStrategy requiredOffloads() {
+                return HttpExecutionStrategies.offloadNone();
+            }
+        };
+    }
+
+    private static class WatchdogHttpLifecycleObserver extends AbstractLifecycleObserverHttpFilter {
+        WatchdogHttpLifecycleObserver(final HttpLifecycleObserver observer, final boolean client) {
+            super(observer, client);
+        }
+    }
+
+    @Override
+    public HttpExchangeObserver onNewExchange() {
+
+        return new HttpExchangeObserver() {
+
+            @Nullable
+            private HttpRequestMetaData requestMetaData;
+
+            @Override
+            public HttpRequestObserver onRequest(final HttpRequestMetaData requestMetaData) {
+                this.requestMetaData = requestMetaData;
+                return NoopHttpLifecycleObserver.NoopHttpRequestObserver.INSTANCE;
+            }
+
+            @Override
+            public HttpResponseObserver onResponse(final HttpResponseMetaData responseMetaData) {
+                return NoopHttpLifecycleObserver.NoopHttpResponseObserver.INSTANCE;
+            }
+
+            @Override
+            public void onExchangeFinally() {
+                if (requestMetaData != null) {
+                    final ContextMap requestContext = requestMetaData.context();
+                    if (requestContext.get(payloadSubscribedKey) == null) {
+                        // No-one subscribed to the payload (or there is none), so if there is a payload
+                        // proactively clean it up.
+                        Publisher<?> payload = requestContext.get(payloadPublisherKey);
+                        if (payload != null) {
+                            LOGGER.debug("Proactively cleaning up HTTP response payload which has been dropped - " +
+                                            "this is a strong indication of a bug in a filter. Request: {}",
```

Comment:
I would probably use `warn` here or maybe even `error`. With `debug`, nobody will see it. 
However, there is a risk to flood logs if we do it for every request. Consider logging only once, by remembering the "logged" flag in a static variable. No need to use volatile, it's ok it if logs a few more times before it becomes visible to other threads. 

Also, consider enhancing the message to clarify that it's in a custom user-defined filter, to make sure they don't think about ST filters.